src/OpenTelemetry.Api/Trace/TracerProvider.cs

```diff
@@ -44,5 +51,92 @@ protected TracerProvider()
     /// <param name="version">Version of the instrumentation library.</param>
     /// <returns>Tracer instance.</returns>
     public Tracer GetTracer(string name, string? version = null)
-        => new(new ActivitySource(name ?? string.Empty, version));
+    {
+        var tracers = this.tracers;
+        if (tracers == null)
+        {
+            // Note: Returns a no-op Tracer once dispose has been called.
+            return new(activitySource: null);
+        }
+
+        var key = threadTracerKey == null
+            ? (threadTracerKey = new TracerKey(name, version))
+            : threadTracerKey.Update(name, version);
+
+        if (tracers[key] is not Tracer tracer)
+        {
+            lock (tracers)
+            {
+                tracer = (tracers[key] as Tracer)!;
+                if (tracer == null)
```

Comment:
This seems a bit odd. `tracer = (tracers[key] as Tracer)!;` is annotated to be not null and then in the `if` check that follows we check for `tracer == null`.

Could we use the same `if` condition here instead (as the one above the lock)?

```suggestion
                if (tracers[key] is not Tracer tracer)
```

Reply:
It is odd, ya. Kind of a quirk caused by the first `is not Tracer tracer` which results in the local `tracer` being tracked as non-null. But the second check may be `null`.

If we do what you are suggesting we'll end up with:

`Error CS0136: A local or parameter named 'tracer' cannot be declared in this scope because that name is used in an enclosing local scope to define a local or parameter`

You could really write it any number of different ways. Here it is without pattern matching on the first read:

```csharp
        var tracer = tracers[key] as Tracer;
        if (tracer == null)
        {
            lock (tracers)
            {
                tracer = tracers[key] as Tracer;
                if (tracer == null)
                {
                    tracer = new(new(key.Name!, key.Version));
                }
            }
        }

        return tracer;
```

If you like that better I'll change it. But I might remove the `Hashtable` completely 😄